File: schedule_backend/app/schemas/sync.py
from __future__ import annotations

from enum import Enum

from pydantic import Field, field_validator

from app.schemas.common import BaseSchema


class SyncOperationEnum(str, Enum):
    create = "create"
    update = "update"
    upsert = "upsert"
    delete = "delete"


class SyncConflictStatusEnum(str, Enum):
    open = "open"
    resolved = "resolved"
    ignored = "ignored"


class SyncStatusRead(BaseSchema):
    enabled: bool
    configured: bool
    logged_in: bool
    server_url: str | None = None
    user_email: str | None = None
    device_id: str
    device_name: str
    pending_count: int
    conflict_count: int
    last_push_at: str | None = None
    last_pull_at: str | None = None
    last_error: str | None = None


class SyncConfigUpdate(BaseSchema):
    server_url: str | None = Field(default=None, max_length=500)
    device_name: str | None = Field(default=None, min_length=1, max_length=100)
    enabled: bool = False

    @field_validator("server_url")
    @classmethod
    def normalize_server_url(cls, value: str | None) -> str | None:
        if value is None:
            return value
        value = value.strip().rstrip("/")
        return value or None


class SyncLoginRequest(BaseSchema):
    email: str = Field(min_length=3, max_length=255)
    password: str = Field(min_length=1, max_length=500)
    server_url: str | None = Field(default=None, max_length=500)
    device_name: str | None = Field(default=None, min_length=1, max_length=100)

    @field_validator("server_url")
    @classmethod
    def normalize_server_url(cls, value: str | None) -> str | None:
        if value is None:
            return value
        value = value.strip().rstrip("/")
        return value or None


class SyncRegisterRequest(BaseSchema):
    email: str = Field(min_length=3, max_length=255)
    password: str = Field(min_length=8, max_length=500)
    display_name: str | None = Field(default=None, max_length=100)
    server_url: str | None = Field(default=None, max_length=500)
    device_name: str | None = Field(default=None, min_length=1, max_length=100)

    @field_validator("server_url")
    @classmethod
    def normalize_server_url(cls, value: str | None) -> str | None:
        if value is None:
            return value
        value = value.strip().rstrip("/")
        return value or None


class SyncRunResult(BaseSchema):
    push_count: int = 0
    pull_count: int = 0
    conflict_count: int = 0
    error_count: int = 0
    message: str = "success"


class SyncConflictRead(BaseSchema):
    id: int
    entity_type: str
    entity_id: str
    status: SyncConflictStatusEnum
    local_payload_json: str | None = None
    remote_payload_json: str | None = None
    created_at: str
    updated_at: str


class SyncConflictListData(BaseSchema):
    items: list[SyncConflictRead]


class SyncConflictResolveRequest(BaseSchema):
    resolution: str = Field(min_length=1, max_length=50)


class SyncConfigRead(BaseSchema):
    status: SyncStatusRead

File: schedule_frontend/src/types/sync.ts
export interface SyncStatus {
  enabled: boolean
  configured: boolean
  logged_in: boolean
  server_url: string | null
  user_email: string | null
  device_id: string
  device_name: string
  pending_count: number
  conflict_count: number
  last_push_at: string | null
  last_pull_at: string | null
  last_error: string | null
}

export interface SyncConfigPayload {
  server_url: string | null
  device_name: string | null
  enabled: boolean
}

export interface SyncConfigResponse {
  status: SyncStatus
}

export interface SyncLoginPayload {
  email: string
  password: string
  server_url?: string | null
  device_name?: string | null
}

export interface SyncRegisterPayload extends SyncLoginPayload {
  display_name?: string | null
}

export interface SyncRunResult {
  push_count: number
  pull_count: number
  conflict_count: number
  error_count: number
  message: string
}

export interface SyncConflict {
  id: number
  entity_type: string
  entity_id: string
  status: 'open' | 'resolved' | 'ignored'
  local_payload_json: string | null
  remote_payload_json: string | null
  created_at: string
  updated_at: string
}

export interface SyncConflictListData {
  items: SyncConflict[]
}

export interface SyncConflictResolvePayload {
  resolution: string
}
